tfx/experimental/pipeline_testing/examples/chicago_taxi_pipeline/taxi_pipeline_regression_e2e_test.py

```diff
@@ -0,0 +1,142 @@
+    """Recursively comparing contents of two directories."""
+
+    dir_cmp = filecmp.dircmp(dir1, dir2)
+    self.assertEmpty(dir_cmp.left_only)
+    self.assertEmpty(dir_cmp.right_only)
+    self.assertEmpty(dir_cmp.funny_files)
+
+    _, mismatch, errors = filecmp.cmpfiles(
+        dir1, dir2, dir_cmp.common_files, shallow=False)
+    self.assertEmpty(mismatch)
+    self.assertEmpty(errors)
+
+    for common_dir in dir_cmp.common_dirs:
+      new_dir1 = os.path.join(dir1, common_dir)
+      new_dir2 = os.path.join(dir2, common_dir)
+      self.assertDirectoryEqual(new_dir1, new_dir2)
+
+  def testTaxiPipelineBeam(self):
+    # Runs the pipeline and record to self._output_dir
+    record_taxi_pipeline = taxi_pipeline_beam._create_pipeline(  # pylint:disable=protected-access
+        pipeline_name=self._pipeline_name,
+        data_root=self._data_root,
+        module_file=self._module_file,
+        serving_model_dir=self._serving_model_dir,
+        pipeline_root=self._pipeline_root,
+        metadata_path=self._metadata_db_uri,
+        beam_pipeline_args=[])
+    BeamDagRunner().run(record_taxi_pipeline)
+    pipeline_recorder_utils.record_pipeline(
+        output_dir=self._output_dir,
+        metadata_db_uri=self._metadata_db_uri,
+        host=None,
+        port=None,
+        pipeline_name=self._pipeline_name,
+        run_id=None)
+
+    # Run pipeline with stub executors.
+    taxi_pipeline = taxi_pipeline_beam._create_pipeline(  # pylint:disable=protected-access, unexpected-keyword-arg
+        pipeline_name=self._pipeline_name,
+        data_root=self._data_root,
+        module_file=self._module_file,
+        serving_model_dir=self._serving_model_dir,
+        pipeline_root=self._pipeline_root,
+        metadata_path=self._metadata_path,
+        beam_pipeline_args=[])
+
+    model_resolver_id = 'ResolverNode.latest_blessed_model_resolver'
+    stubbed_component_ids = [component.id
+                             for component in taxi_pipeline.components
+                             if component.id != model_resolver_id]
+
+    stub_launcher = stub_component_launcher.get_stub_launcher_class(
+        test_data_dir=self._output_dir,
+        stubbed_component_ids=stubbed_component_ids,
+        stubbed_component_map={})
+    stub_pipeline_config = pipeline_config.PipelineConfig(
+        supported_launcher_classes=[
+            stub_launcher,
+        ])
+    BeamDagRunner(config=stub_pipeline_config).run(taxi_pipeline)
+
+    self.assertTrue(tf.io.gfile.exists(self._metadata_path))
+
+    metadata_config = metadata.sqlite_metadata_connection_config(
+        self._metadata_path)
+
+    # Verify that recorded files are successfully copied to the output uris.
+    with metadata.Metadata(metadata_config) as m:
+      artifacts = m.store.get_artifacts()
+      artifact_count = len(artifacts)
+      execution_count = len(m.store.get_executions())
+      self.assertGreaterEqual(artifact_count, execution_count)
```

Comment:
Is there a reason that number of artifact should greater or equal to the number of executions? Then we might have some comments about why.